### ISSUE TITLE
Prevent adding high content-length to the header

### ISSUE THREAD

**Issue by suzanaAngelovaIvanova** [eclipse/hono#3263]
When we try to execute a request (curl -H "content-length:1844674407370" -i -X POST http://host:port/v1/tenants) and add a content-length to the request that is larger from the max body size and if it doesn't **equal** to the real body size, the request cannot be processed and returns a timeout.
To prevent it, we will create a handler to validate the content-length and return a bad request if the validation fails.

**Comment by sophokles73** [comment]
This actually sounds like a check that should be implemented in vert.x, doesn't it?

**Comment by suzanaAngelovaIvanova** [comment]
> This actually sounds like a check that should be implemented in vert.x, doesn't it?

Yes, you are right. In fact, the problem is reproduced when we put manually an unreal/wrong content-length in the header.

The method setBodyLimit in vert.x is used when we have a request with real content-length and we set a body size limit, 
as content-length is calculated automatically when sending the request and if content-length > limit body, we will get error 413.
Our test case is different, if someone decides to add content-length manually and the content-length is not calculated automatically when sending the request, respectively the content-length is unreal/wrong. In this case the request cannot be processed and returns a timeout.
Therefore, validation is needed, the validation can be done with a new handler and the handler can be added in the first place in the chain.

**Comment by calohmn** [comment]
A change has been done in vert-web 4.3.0 that checks the content-length header early on:
```
      // before parsing the body we can already discard a bad request just by inspecting the content-length against
      // the body limit, this will reduce load, on the server by totally skipping parsing the request body
```
See https://github.com/vert-x3/vertx-web/commit/7ab8cee7c6b863a7c55ae53d69274b8a4b5f730a.
There is no Quarkus version with vert.x 4.3.0 yet, though.

**Comment by calohmn** [comment]
I think we should also consider the more general case here:
Currently, when the client is doing a POST/PUT request with, let's say, `content-length: 10` but not sending any body data at all, just keeping the request open, the server will wait for data for a long time (in my test it was 60s). Such a request could be obviously invalid, because it for example lacks authentication data, but the server would still spend resources on it for quite a long time.

Therefore, I think we should cancel client requests that are invalid early on, before parsing the body.
Currently, in `HttpServiceBase.createRouter()` and `AbstractVertxBasedHttpProtocolAdapter.createRouter()`, the `"BodyHandler with request size limit"` gets added as one of the first items, before the Auth handlers. Can this be changed so that the Auth handlers come first?
I guess all requests not matching any of the paths of the specific routes (like `telemetry`, `event`, `command` in the HTTP adapter) could be responded to early on as well, not waiting for the HTTP body.

**Comment by avgustinmm** [comment]
Actually, hono (or vertx-web 4.3+) might discard content lengths bigger then max size (even before reading)
but can't discard allowed (less then limit) but wrong (e.g. intentionally wrong) content length bigger then real payload.
In that case vertx will expect to read more data until read timeout occur. By default it seems that default idle read timeout is 0 and hono will wait to read data forever keeping connection (and associated resources) open. 
Probably reasonable read timeouts shall be set in HttpServerOptions when creating https servers. The same applies for Http adapters.

**Comment by sophokles73** [comment]
> Can this be changed so that the Auth handlers come first?

+1 for adding the BodyHandler after the Auth handler and to routes with a specific URI only (i.e. not to the match-all-route).

**Comment by calohmn** [comment]
> By default it seems that default idle read timeout is 0 and hono will wait to read data forever keeping connection (and associated resources) open.
Probably reasonable read timeouts shall be set in HttpServerOptions when creating https servers. 

Yes, I agree on setting a reasonable default here. 
I think, the timeout should also be configurable. The timeout configuration could be added to the `HONO_HTTP_*` [HTTP adapter properties](https://www.eclipse.org/hono/docs/dev/admin-guide/http-adapter-config/#service-configuration) and `HONO_REGISTRY_HTTP_*` device registry HTTP server properties.
If not defined there, we could also apply the corresponding [quarkus.http.*](https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus-vertx-http-eclipse-vert.x-http) properties, like it is done [here](https://github.com/quarkusio/quarkus/blob/18d34eb5b22f43fedba7ed0b8327a9a03f54507e/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java#L770). I guess these property values are also applied to the Smallrye Health HTTP server in Hono components.

**Comment by sophokles73** [comment]
> If not defined there, we could also apply the corresponding [quarkus.http.*](https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus-vertx-http-eclipse-vert.x-http) properties, like it is done [here](https://github.com/quarkusio/quarkus/blob/18d34eb5b22f43fedba7ed0b8327a9a03f54507e/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java#L770). I guess these property values are also applied to the Smallrye Health HTTP server in Hono components

Actually, the values set using the *quarkus.http.* properties will **only** be applied to the HTTP server started for serving the metrics and health checks because that is the HTTP server managed by Quarkus.

In order to configure the HTTP server used for processing device requests in the HTTP adapter and in the Device Registry, we need to use the *HONO_HTTP_* properties, because we are starting a separate HTTP server there.

**Comment by calohmn** [comment]
> Actually, the values set using the *quarkus.http.* properties will **only** be applied to the HTTP server started for serving the metrics and health checks because that is the HTTP server managed by Quarkus.

Sure. Yes, thinking more about it, although re-using the existing Quarkus server configuration possibilities for the Hono managed HTTP server would be convenient, it's probably better to keep the configurations separate, as also the use cases for these servers are different. I guess adding `hono.http.idleTimeout` (HTTP adapter) and `hono.registry.http.idleTimeout` (device registry) would be sufficient then.

**Comment by sophokles73** [comment]
@suzanaAngelovaIvanova @calohmn
Is there anything left that needs to be done on branches 2.0.x and 1.12.x to fix this?

**Comment by sophokles73** [comment]
Closed via #3313
